apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/EventReportServiceClient.java

```diff
@@ -0,0 +1,178 @@
+                             .setSource(
+                                 Source.newBuilder()
+                                       .setService(Config.Agent.SERVICE_NAME)
+                                       .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                                       .build()
+                             )
+                             .putParameters(
+                                 "OPTS",
+                                 runtimeMxBean.getInputArguments()
+                                              .stream()
+                                              .sorted()
+                                              .collect(Collectors.joining(" "))
+                             );
+    }
+
+    @Override
+    public void boot() throws Throwable {
+
+    }
+
+    @Override
+    public void onComplete() throws Throwable {
+        reportStartingEvent();
+    }
+
+    @Override
+    public void shutdown() throws Throwable {
+        if (!CONNECTED.equals(status)) {
+            return;
+        }
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        final Event.Builder shutdownEvent = Event.newBuilder()
+                                                 .setUuid(UUID.randomUUID().toString())
+                                                 .setName("Shutdown")
+                                                 .setStartTime(System.currentTimeMillis())
+                                                 .setEndTime(System.currentTimeMillis())
+                                                 .setMessage("Shutting down Java Application")
+                                                 .setType(Type.Normal)
+                                                 .setSource(
+                                                     Source.newBuilder()
+                                                           .setService(Config.Agent.SERVICE_NAME)
+                                                           .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                                                           .build()
+                                                 );
+
+        final StreamObserver<Event> collector = eventServiceStub.collect(new StreamObserver<Commands>() {
+            @Override
+            public void onNext(final Commands commands) {
+                ServiceManager.INSTANCE.findService(CommandService.class).receiveCommand(commands);
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                LOGGER.error("Failed to report shutdown event.", t);
+                // Ignore status change at shutting down stage.
+                latch.countDown();
+            }
+
+            @Override
+            public void onCompleted() {
+                latch.countDown();
+            }
+        });
+
+        collector.onNext(shutdownEvent.build());
+        collector.onCompleted();
+        latch.await();
```

Comment:
Consider using a timeout mechanism. This way could block the shutdown process.

Reply:
> Consider using a timeout mechanism. This way could block the shutdown process.

No, the gRPC service would either success (`onCompleted` called) or timeout (`onError` called), either of them should call `latch.countDown()` and `latch.await` won't block the shutdown process

Reply:
I can't see you are using `withDeadlineAfter`, take a look at `TraceSegmentServiceClient` about how to do timeout in gRPC.